### PR TITLE
Fix duplicate department IDs

### PR DESCRIPTION
**Pull request by blakekrammes**
- Create new dept ids at xxx11 and up to avoid colliding with historic ids in intacct
- Select randomized new dept id from list of unused ids in the acceptable range –– this will prevent duplicate dept ids stemming from simultaneous requests in the vast majority of cases

### CODE REVIEW

File: src/components/project/handlers/set-department-id.handler.ts [hunk 1]
```diff
@@ -46,19 +46,18 @@ export class SetDepartmentId implements IEventHandler<SubscribedEvent> {
...
+        WITH "${departmentIdPrefix}" + substring("0000", 1, 4 - size(nextIdBase)) + nextIdBase AS nextId
+        MATCH (project:Project {id: "${project.id}"})
         OPTIONAL MATCH (project)-[oldDepartmentIdRelationship:departmentId {active: true}]-(oldDepartmentIdPropertyNode:Property)
         SET oldDepartmentIdRelationship.active = false
         WITH coalesce(oldDepartmentIdPropertyNode.value, nextId) as departmentId, project
         CREATE (project)-[newDepartmentIdRelationship:departmentId { active: true, createdAt: datetime() }]->(:Property { createdAt: datetime(), value: departmentId })
         RETURN departmentId
-        `,
-        { departmentIdPrefix: departmentIdPrefix, projectId: project.id }
+        `
```
Comment (CarsonF) [inline src/components/project/handlers/set-department-id.handler.ts:60]: This is just concatenating the parameters into the query which is not what we want to do. We want parameters to be bound for safety & performance.

Parameters are only bound when using [`raw`](https://jamesfer.me/cypher-query-builder/classes/query.html#raw) as a template tag. i.e.
```ts
raw`match ... ${param}`
```
instead of a function call
```ts
raw(`match ${param}`) // <- string concatenation
```

**Review by CarsonF — CHANGES_REQUESTED** [review]

File: src/components/project/handlers/set-department-id.handler.ts (hunk 1 above)
Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts:60]: What does binding them do? How is it different?

Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts]: So it's easier for the engine to parse?

Comment (CarsonF) [reply src/components/project/handlers/set-department-id.handler.ts]: [Yes](https://neo4j.com/docs/cypher-manual/current/syntax/parameters/) but it also prevents
things like `"${foo}"` where foo is something like `"; match (n) detach delete n; "`. I know these variables are not directly user input, but it's just good practice. And better for the database as the link states.

Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts]: Ok. Thanks for the info. Makes sense. At first I thought it was a problem with the query builder variable, but it was just not a string. I fixed the function that grabs the funding account number for this to actually return a string instead of a number.

I think it's ready to go then.

**Comment by blakekrammes** [comment]
I think the problem was that `$departmentIdPrefix` on 49 was a number when it should've been a string. I think this was causing problems matching which dept ids already existed.

**Comment by blakekrammes** [comment]
There's also some TODO work involving fixing projects that have duplicate dept ids.

**Comment by blakekrammes** [comment]
I just tested out using a number in that query and it threw an error. I don't know how that query was going through.

**Comment by blakekrammes** [comment]
Here are screenshots running the test  on develop with the old code and on this branch with the changes:

Screen Shot 2021-03-10 at 10 32 49 AM
Screen Shot 2021-03-10 at 10 42 45 AM

File: src/components/project/handlers/set-department-id.handler.ts [hunk 2]
```diff
@@ -48,9 +48,9 @@ export class SetDepartmentId implements IEventHandler<SubscribedEvent> {
         MATCH ()-[:departmentId]-(departmentIdPropertyNode:Property)
         WHERE departmentIdPropertyNode.value STARTS WITH $departmentIdPrefix
```
Comment (blakekrammes) [inline src/components/project/handlers/set-department-id.handler.ts:49]: Line 49 is where the problem actually was I think. Since `$departmentIdPrefix` was a number and not a string I think it wasn't matching the existing dept ids. However, since the same value was being inserted as a JS template literal var on 53 and surrounded by quotes, that bit was working.

File: test/project.e2e-spec.ts [hunk 3]
```diff
@@ -1140,4 +1140,42 @@ describe('Project e2e', () => {
       projectQueryResult.project.budget.value.records[0].organization.value.id;
     expect(firstBudgetRecordOrganizationId).toBe(org.id);
   });
+
+  it('should not assign the same department id to two projects created with the same location', async () => {
+    await loginAsAdmin(app);
```
[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: should it not be used anywhere?

File: src/components/project/project.service.ts [hunk 4]
```diff
@@ -135,6 +135,8 @@ export class ProjectService {
     return [
       'CREATE CONSTRAINT ON (n:Project) ASSERT EXISTS(n.id)',
       'CREATE CONSTRAINT ON (n:Project) ASSERT n.id IS UNIQUE',
+      // this is not a real constraint since there is currently no DepartmentId label in the db
+      // we may want to add one?
```
[the remark this answers is not in the document]
Comment (blakekrammes) [reply src/components/project/project.service.ts:139]: Separate PR?

File: src/components/project/handlers/set-department-id.handler.ts [hunk 5]
```diff
@@ -72,11 +72,12 @@ export class SetDepartmentId implements IEventHandler<SubscribedEvent> {
     const res = await this.db
       .query()
       .raw(
+        // convert returned value from float to int and then to string
         `
         MATCH (:Project {id: $projectId})-[:primaryLocation {active: true}]
               -()-[:fundingAccount {active: true}]
               -()-[:accountNumber {active: true}]-(node:Property)
-        RETURN node.value as prefix
+        RETURN toString(toInteger(node.value)) as prefix
```
Comment (CarsonF) [inline src/components/project/handlers/set-department-id.handler.ts:80]: What do you think about just changing the as result type to be correct
```tsx
.asResult<{ prefix: number }>()
```
And then just converting it to a string in TS
```tsx
return res.prefix.toString();
```

I would actually kinda like to replace "department id prefix" with "funding account number" for clarity also....

File: test/project.e2e-spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: Also speaking of problems with concurrent requests. I tried `Promise.all`-ing the creation and updating of two projects and the ended up with the same dept id. If done synchronously, as is done in the test now, they don't.

File: src/components/project/handlers/set-department-id.handler.ts (hunk 5 above)
Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts:80]: Yeah either way. I guess it's a bit cleaner. I just feel cooler doing it neo4j. I can change the name too.

Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts:80]: So JS converts floats ending in zero `1.0` automatically into numbers of 1. That's fun.

File: src/components/project/project.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (blakekrammes) [reply src/components/project/project.service.ts:139]: Lol. It's a couple lines I think actually.

[the remark this answers is not in the document]
Comment (CarsonF) [reply src/components/project/project.service.ts:139]: Nah here.

File: src/components/project/handlers/set-department-id.handler.ts (hunk 5 above)
Comment (CarsonF) [reply src/components/project/handlers/set-department-id.handler.ts:80]: Right because `1 === 1.0`. This is why the TS type is `number` instead of `integer` or `float`. JS doesn't have distinct types for these.

File: test/project.e2e-spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (CarsonF) [reply test/project.e2e-spec.ts:1145]: Adding the label with the unique constraint would cause this to fail instead of creating a duplicate. Maybe start with that to confirm the problem.

File: src/components/project/handlers/set-department-id.handler.ts (hunk 5 above)
Comment (blakekrammes) [reply src/components/project/handlers/set-department-id.handler.ts:80]: Right.

File: test/project.e2e-spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: That's what I was hoping. It seems like we should lock all unique properties if we're doing that, not just this one.

[the remark this answers is not in the document]
Comment (CarsonF) [reply test/project.e2e-spec.ts:1145]: Neo4j normally handles the locking internally, and unique constraint violations just throw an error so there's no reason to do anything different normally. This is an edge case where we have complex query logic that neo4j can't figure out implicitly.

[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: When do constraints get applied?

[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: So the constraint is working if I try and change a dept id in the db to the same as another. It isn't working in the test suite. I'm wondering if the constraint isn't getting applied.

[the remark this answers is not in the document]
Comment (CarsonF) [reply test/project.e2e-spec.ts:1145]: Constraints are applied when app starts & db connects. Unless you have
```
DB_CREATE_INDEXES=false
```
in your .env.local

Did you add the label [here](https://github.com/SeedCompany/cord-api-v3/pull/1731/files#diff-d4b319fbc9fb22707d5969d0591543ce306f553ffcadfd4397fc912029545b4eR58)?

[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: So the constraint works, but not for tests. Dunno if it's not getting applied properly. 

You're saying that in this case it's not sufficient to have a unique constraint? Will it leave the db in a partial state?

[the remark this answers is not in the document]
Comment (CarsonF) [reply test/project.e2e-spec.ts:1145]: The resetDatabase call removes all constraints after the test suite, maybe that's some how applying out of order. I wouldn't worry about it too much because 👇 

> You're saying that in this case it's not sufficient to have a unique constraint? Will it leave the db in a partial state?

No it's not sufficient here because the change isn't directly from a user. They can't help if the same dept id was chosen at the same time as another.

Also yes, almost all errors will leave database in a partial state until transactions can be applied.

[the remark this answers is not in the document]
Comment (blakekrammes) [reply test/project.e2e-spec.ts:1145]: I did add the label. I should push my changes so far.

File: src/components/project/handlers/set-department-id.handler.ts [hunk 6]
```diff
@@ -48,9 +46,10 @@ export class SetDepartmentId implements IEventHandler<SubscribedEvent> {
         MATCH ()-[:departmentId]-(departmentIdPropertyNode:Property)
         WHERE departmentIdPropertyNode.value STARTS WITH $departmentIdPrefix
         WITH collect(distinct(toInteger(right(departmentIdPropertyNode.value, 4)))) as listOfDepartmentIds
-        UNWIND [n IN range(1, 9999) WHERE NOT n IN listOfDepartmentIds] as listOfUnusedDepartmentIds
-        WITH toString(min(listOfUnusedDepartmentIds)) AS nextIdBase
-        WITH '${departmentIdPrefix}' + substring("0000", 1, 4 - size(nextIdBase)) + nextIdBase as nextId
+        WITH [n IN range(11, 9999) WHERE NOT n IN listOfDepartmentIds] as listOfUnusedDepartmentIds
```
Comment (blakekrammes) [inline src/components/project/handlers/set-department-id.handler.ts:49]: This worked! Thanks, @CarsonF

File: test/project.e2e-spec.ts [hunk 7]
```diff
@@ -1140,4 +1140,47 @@ describe('Project e2e', () => {
...
+          `,
+          {
+            id: project.id,
+            // updating to this step assigns a dept id
+            step: ProjectStep.PendingFinanceConfirmation,
+          }
+        );
+        return updatedProject.updateProject.project;
+      };
+      const projects = await Promise.all(
```
Comment (blakekrammes) [inline test/project.e2e-spec.ts:1176]: Also note that this is running in parallel now and doesn't produce duplicate dept ids (yet) :)

**Comment by blakekrammes** [comment]
@CarsonF think this is ready to go pending tests.

**Review by CarsonF — APPROVED** [review]
Thanks @blakekrammes! Looks great
